video/decode/vd_lavc.c

```diff
@@ -952,6 +958,40 @@ static enum AVPixelFormat get_format_hwdec(struct AVCodecContext *avctx,
     return select;
 }
 
+static int get_buffer2_direct(AVCodecContext *avctx, AVFrame *pic, int flags)
+{
+    struct dec_video *vd = avctx->opaque;
+    struct vo_buffer_data vbufd = { 0 };
+
+    struct mp_image img;
+    mp_image_copy_fields_from_av_frame(&img, pic);
```

Comment:
I think it would make sense to pass width/height/pixfmt as separate parameter, and then let the callback return an mp_image. Then you'd convert it to an AVFrame reference, and move it to the target "pic". Not sure if that works out, but we wouldn't need any additional tricky things.